nursery/hash-data-using-ripemd128.yml

```diff
@@ -0,0 +1,26 @@
+rule:
+  meta:
+    name: hash data using RIPEMD128
+    namespace: data-manipulation/hashing/ripemd128
+    authors:
+      - raymond.leong@mandiant.com
+    scope: function
+    references:
+      - https://en.wikipedia.org/wiki/RIPEMD-128
+  features:
+    - and:
+      - or:
+        - and:
+          - description: Magic initialization constants used in RIPEMD-128
+          - number: 0x67452301 = also used in MD5, RIPEMD-128, RIPEMD-256, and RIPEMD-320
+          - number: 0xEFCDAB89 = also used in MD5, RIPEMD-128, RIPEMD-256, and RIPEMD-320
+          - number: 0x98BADCFE = also used in MD5, RIPEMD-128, RIPEMD-256, and RIPEMD-320
+          - number: 0x10325476 = also used in MD5, RIPEMD-128, RIPEMD-256, and RIPEMD-320
+        - bytes: 01 23 45 67 89 AB CD EF FE DC BA 98 76 54 32 10
+
+      - not:
+        - or:
+          - number: 0x76543210 = RIPEMD-256 and RIPEMD-320 constant
+          - number: 0xFEDCBA98 = RIPEMD-256 and RIPEMD-320 constant
+          - number: 0xC3D2E1F0 = RIPEMD-160 and RIPEMD-320 constant
+          - number: 0x3C2D1E0F = RIPEMD-320 constant
```

Comment:
is there a way to exclude MD5 here?

Reply:
i don't totally understand all the cases here and if you've covered them, but i trust you to have done that work :-)

Reply:
Good point and added an additional function scope for an array constant not present in MD5. Would have liked to keep it a function scoped rule but the magic constants are sometimes in a function that solely moves their values into a structure with nothing else to signature off:
<img width="481" alt="Screen Shot 2023-02-14 at 11 47 18 PM" src="https://user-images.githubusercontent.com/77110701/219266655-a8c3c377-a0c9-466e-8dcb-c3eba7a87538.png">

Reply:
The RIPEMD160 rule is having a similar issue with SHA1. The magic constants listed in https://rst.im/magic/RIPEMD160 are also present in SHA1, including the K constants even though they arent listed in the corresponding SHA1 page. The routines have a lot of similarities and more research is needed to write a rule that better deconflicts the two. I think keeping the RIPEMD160 rule out of the repo may be better for now to avoid false positives. 

Of note, the logic in the SHA1 rule below will also hit on RIPEMD initialization but with SHA1 being much more common than the other two may be less of an issue:
```
- or:
      - and:
        - description: Magic initialization constants used in SHA1
        - number: 0x67452301 = A, also used by MD5
        - number: 0xEFCDAB89 = B, also used by MD5
        - number: 0x98BADCFE = C, also used by MD5
        - number: 0x10325476 = D, also used by MD5
        - number: 0xC3D2E1F0 = specific to SHA1, not MD4 nor MD5
```

Reply:
awesome, thanks for the explanation. it seems a little tricky to get perfectly right, but i tend to agree with you that we should focus on the mostly likely scenarios and being helpful to most analysts. maybe disabling one of the rules and adding extra documentation the other is best?

Reply:
Thanks for the suggestion. Updated the documentation in the MD5 and SHA1 rules to highlight the shared constants with the RIPEMD routines and removed the RIPEMD160 rule that would be prone to false positives with SHA1 (didnt see a way to disable it).